File: 0042-trapping-rain-water/solution.java
class Solution {
    public int trap(int[] h) {
        int n = h.length;
        int[] left = new int[n];
        int[] right = new int[n];
        
        int max = h[0];
        left[0] = h[0];
        for(int i=1;i<n;i++){
            max = Math.max(max,h[i]);
            left[i] = max;
        }

        int maxR = h[n-1];
        right[n-1] = h[n-1];
        for(int i = n-2;i>=0 ;i--){
           maxR = Math.max(maxR,h[i]);
            right[i] = maxR;
        }

        int[] ans = new int[n];
        int sum = 0;
        for(int i=0;i<n;i++){
            int min = Math.min(left[i],right[i]);
            ans[i] = min;
            sum += min;
        }

        int sum1 = 0;
        for(int i : h){
            sum1 += i;
        }
        return sum-sum1;
    }
}

File: 0042-trapping-rain-water/solution.c
int trap(int* h, int n) {
    int left[n];
    int right[n];
    int ans[n];
    int max=h[0],max1=h[n-1];
    for(int i=0;i<n;i++){
        max = fmax(h[i],max);
        left[i] = max;
    }
    for(int i=n-1;i>=0;i--){
        max1 = fmax(h[i],max1);
        right[i]=max1;
    }
    int sum=0;
    for(int i=0;i<n;i++){
        max = fmin(left[i],right[i]);
        ans[i]=max;
        sum+=ans[i];
    }
    int sum1=0;
    for(int i=0;i<n;i++){
        sum1+=h[i];
    }
    printf("%d %d",sum,sum1);
    return sum-sum1;
}
